File: 2527-count-subarrays-with-fixed-bounds/2527-count-subarrays-with-fixed-bounds.cpp
#pragma GCC optimize("03", "unroll-loops");

static const int __ = [](){
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    return 0;
}();

class Solution {
public:
    long long countSubarrays(vector<int>& nums, int minK, int maxK) {
        long long ans = 0;
        int invalid_idx = -1, min_idx = -1, max_idx = -1;

        for (int i = 0; i < nums.size(); ++i) {
            if (!(minK <= nums[i] && nums[i] <= maxK)) {
                invalid_idx = i;
            }

            if (nums[i] == minK) {
                min_idx = i;
            }

            if (nums[i] == maxK) {
                max_idx = i;
            }

            ans += max(0, min(min_idx, max_idx) - invalid_idx);
        }

        return ans;
    }
};

File: 2527-count-subarrays-with-fixed-bounds/2527-count-subarrays-with-fixed-bounds.go
func countSubarrays(nums []int, minK int, maxK int) int64 {
    ans := int64(0)
    lastmK, lastMk, lastInvalid := -1, -1, -1
    for i, num := range nums {
        if num < minK || num > maxK {
            lastInvalid = i
        }
        if num == minK {
            lastmK = i
        }
        if num == maxK {
            lastMk = i
        }

        validStart := min(lastmK, lastMk)
        if validStart > lastInvalid {
            ans += int64(validStart - lastInvalid)
        }
    }
    return ans
}